### PR TITLE
implement range/generate_series func

### LINKED ISSUE #8028: Support udf `range`
### Is your feature request related to a problem or challenge?

Currently our udf function does not have `range`. This is a commonly used function. I understand that I should add it.
try:
`select range(10);`
result: 
`Error: Plan("Invalid function 'range'.\nDid you mean 'nanvl'?")`
 
And duckdb or clickhouse has supported it.
duckdb: [https://duckdb.org/docs/archive/0.9.1/sql/functions/](https://duckdb.org/docs/archive/0.9.1/sql/functions/nested#:~:text=%27k%27-,Range%20Functions,-The%20functions%20range)
clickhouse: [https://clickhouse.com/docs/en/sql-reference/functions/array-functions](https://clickhouse.com/docs/en/sql-reference/functions/array-functions#rangeend-rangestart--end--step:~:text=Syntax-,range(%5Bstart%2C%20%5D%20end%20%5B%2C%20step%5D),-Arguments)

### Describe the solution you'd like

Add the udf function.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Veeupup**
## Which issue does this PR close?

Closes #8028 .

## Are these changes tested?

test in sqllogictests

### CODE REVIEW

**Comment by Veeupup** [comment]
cc @jayzhan211  @alamb  this PR is ready for review! : )

**Comment by 2010YOUY01** [comment]
Thank you for this new function! Looks like these functions are available in DuckDB 👍🏼 https://duckdb.org/docs/sql/functions/nested.html#range-functions

Here are some suggestions:
1. Can we add more corner case tests like `range(5,5,0), range(1,5,-1), range(1,-5,1)...`
2. Include the semantics of this function (is start/end index inclusive or exclusive, what will happen for `range(1,5,-1)`) in SQL doc: https://github.com/apache/arrow-datafusion/blob/main/docs/source/user-guide/sql/scalar_functions.md

File: datafusion/expr/src/expr_fn.rs [hunk 1]
```diff
@@ -722,6 +722,12 @@ scalar_expr!(
     "Returns an array of the elements in the intersection of array1 and array2."
 );
 
+nary_scalar_expr!(
```
Comment (jayzhan211) [inline datafusion/expr/src/expr_fn.rs:725]: Should this be `scalar_expr`?

File: datafusion/expr/src/expr_fn.rs [hunk 2]
```diff
@@ -722,6 +722,12 @@ scalar_expr!(
...
+    Range,
+    gen_range,
```
Comment (jayzhan211) [inline datafusion/expr/src/expr_fn.rs:727]: start, stop, step

**Review by jayzhan211 — APPROVED** [review]
LGTM!

File: datafusion/expr/src/expr_fn.rs (hunk 1 above)
Comment (Veeupup) [reply datafusion/expr/src/expr_fn.rs]: fixed! thanks !

**Comment by Veeupup** [comment]
> Thank you for this new function! Looks like these functions are available in DuckDB 👍🏼 https://duckdb.org/docs/sql/functions/nested.html#range-functions
> 
> Here are some suggestions:
> 
> 1. Can we add more corner case tests like `range(5,5,0), range(1,5,-1), range(1,-5,1)...`
> 2. Include the semantics of this function (is start/end index inclusive or exclusive, what will happen for `range(1,5,-1)`) in SQL doc: https://github.com/apache/arrow-datafusion/blob/main/docs/source/user-guide/sql/scalar_functions.md

thanks! I have made it clear in the test and doc.

File: docs/source/user-guide/sql/scalar_functions.md [hunk 3]
```diff
@@ -2426,6 +2427,20 @@ trim_array(array, n)
   Can be a constant, column, or function, and any combination of array operators.
 - **n**: Element to trim the array.
 
+### `range`
```
Comment (2010YOUY01) [inline docs/source/user-guide/sql/scalar_functions.md:2430]: This is clear and precise 💯

File: datafusion/physical-expr/src/array_expressions.rs [hunk 4]
```diff
@@ -579,6 +579,41 @@ pub fn array_pop_back(args: &[ArrayRef]) -> Result<ArrayRef> {
...
+        ),
+        _ => return internal_err!("gen_range expects 1 to 3 arguments"),
+    };
+
+    let mut values = vec![];
+    let mut offsets = vec![0];
+    for (idx, stop) in stop_array.iter().enumerate() {
+        let stop = stop.unwrap_or(0);
+        let start = start_array.as_ref().map(|arr| arr.value(idx)).unwrap_or(0);
+        let step = step_array.as_ref().map(|arr| arr.value(idx)).unwrap_or(1);
```
Comment (2010YOUY01) [inline datafusion/physical-expr/src/array_expressions.rs:603]: We might not want the bad input to crash at run time
```
❯ select range(1,5,0);
thread 'main' panicked at 'assertion failed: step != 0', /rustc/e6d4725c76f3b526c74454bc51afdf6daf133506/library/core/src/iter/adapters/step_by.rs:21:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
Maybe `if step == 0 { return exec_err!("step can't be 0 for function range(start [, stop, step]")}`?

**Review by 2010YOUY01 — COMMENTED** [review]
Thank you for the doc. I think the panic should be addressed. Everything else looks good to me

File: datafusion/physical-expr/src/array_expressions.rs (hunk 4 above)
Comment (Veeupup) [reply datafusion/physical-expr/src/array_expressions.rs:603]: thanks for your advice! fixed!  : )

File: datafusion/physical-expr/src/array_expressions.rs [hunk 5]
```diff
@@ -579,6 +579,44 @@ pub fn array_pop_back(args: &[ArrayRef]) -> Result<ArrayRef> {
     )
 }
 
+pub fn gen_range(args: &[ArrayRef]) -> Result<ArrayRef> {
```
Comment (alamb) [inline datafusion/physical-expr/src/array_expressions.rs:582]: This is a very nice implementation 👨‍🍳 👌 

Can you please add a docstring that explains that the arguments are and how they are interpreted

Perhaps following the great example of @jayzhan211  in https://github.com/apache/arrow-datafusion/blob/93a95775f51d21445958067b1da0991879464bb9/datafusion/physical-expr/src/array_expressions.rs#L582-L600

**Review by alamb — APPROVED** [review]
Thank you @Veeupup  -- this looks great. I took the liberty of merging this branch up from main to resolve a conflict. 

It would be great if you could add a doc comment explaining the arguments to `gen_range` , but I also think we could do that as a follow on PR if necessary.

Thanks again for the recent string of nice PRs 🙏

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Thanks again @Veeupup and @jayzhan211
